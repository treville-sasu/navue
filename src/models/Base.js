export class Model {
  // eslint-disable-next-line no-unused-vars
  constructor({ type, ...properties } = {}) {
    Object.assign(this, properties);
  }

  toJSON() {
    return {
      ...this,
      type: this.constructor.name
    };
  }

  // TODO write specs
  clone() {
    return this.constructor.from(JSON.parse(JSON.stringify(this)));
  }

  static from({ type, ...properties } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) return new this(properties);
    else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }
}

// TODO : implementation for when Vuejs Suppots Array-like objects.
// export class Store extends Array {
export class Store {
  constructor(properties = {}, ...items) {
    this.items = items;
    Object.assign(this, properties);
  }

  // get items() {
  //   return [...this];
  // }

  [Symbol.iterator]() {
    return this.items[Symbol.iterator]();
  }

  get length() {
    return this.items.length;
  }

  valueOf() {
    return this.items;
  }

  get properties() {
    // eslint-disable-next-line no-unused-vars
    let { items, ...props } = { ...this };
    return props;
    // return Object.fromEntries(
    //   Object.entries(this).filter(([key]) => isNaN(key))
    // );
  }

  get first() {
    return this.items[0];
  }

  get last() {
    return this.items[this.length - 1];
  }

  add(item, index) {
    index ? this.items.splice(index, 0, item) : this.items.push(item);
    return this.length;
  }

  remove(item, index) {
    if (index === undefined) index = this.items.indexOf(item);
    return index >= 0 ? this.items.splice(index, 1).length == 1 : false;
  }

  keep(filter) {
    let i = this.items.length;
    while (i--) {
      let result = filter.call(this.items, this.items[i]);

      if (result == undefined) this.items.splice(i, 1);
      else if (!Object.is(result, this.items[i]))
        this.items.splice(i, 1, result);
    }
    return this;
  }

  toJSON() {
    return {
      ...this.properties,
      items: this.items,
      type: this.constructor.name
    };
  }

  static from({ type, items = [], ...properties }, ...constructors) {
    if (type != this.name)
      throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
    else {
      let constructor = constructors.shift();
      return new this(
        properties,
        ...items.map(i => {
          return constructor ? constructor.from(i, ...constructors) : i;
        })
      );
    }
  }
}
