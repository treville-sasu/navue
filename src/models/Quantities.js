import { Model } from "@/models/Base.js";

export class Quantity extends Model {
  // eslint-disable-next-line no-unused-vars
  constructor(value, _unit, { type, unit, precision, ...properties } = {}) {
    if (arguments[0] instanceof Quantity) return arguments[0];
    else {
      super(properties);

      if (_unit) this.unit = _unit;
      else this._unit = undefined;

      if (value || value === 0) this.value = value;
      else {
        this._value = undefined;
        this.precision = undefined;
      }

      if (precision) this.precision = precision;

      if (unit) this.unit = unit;
    }
  }

  get value() {
    return this.constructor.significantValue(
      this._to(this._value),
      this.precision
    );
  }

  set value(val) {
    let f = Number(val);
    if (Number.isNaN(f)) throw `\`${val}\` could not be coerced to a number`;
    else {
      this._value = this._from(f);
      this.precision = Math.max(3, this.constructor.significantFigures(val));
    }
  }

  get unit() {
    return this._unit || this.constructor.baseUnit;
  }

  set unit(val) {
    if (!this.constructor.units[val]) throw `'${val}' is not an available unit`;
    else this._unit = val;
  }

  get factor() {
    return this.constructor.units[this.unit];
  }

  _from(val, unit = this.unit) {
    return Number(val) / this._factor(unit);
  }

  _to(val, unit = this.unit) {
    return Number(val) * this._factor(unit);
  }

  _factor(unit) {
    if (this.constructor.units[unit]) return this.constructor.units[unit];
    else throw `'${unit}' is not an available unit`;
  }

  to(unit) {
    return this._to(this.value, unit);
  }

  as(unit, precision) {
    return Object.assign(this.constructor.from(this), { unit, precision });
  }

  add(other, props) {
    if (typeof other == "number" || other instanceof this.constructor)
      return new this.constructor(this + other, undefined, {
        precision: Math.min(this.precision, other.precision),
        unit: this._unit,
        ...props
      });
    else throw "operands should be of the same type";
  }

  // FIXME : prod with quantity should not return the type from this nor other.
  prod(other, props) {
    return new this.constructor(this * other, undefined, {
      precision: Math.min(
        this.precision,
        other.precision || this.constructor.significantFigures(other)
      ),
      unit: this._unit,
      ...props
    });
  }

  valueOf() {
    return this._value;
  }

  toString() {
    let sf = Math.max(1, Math.min(this.precision, 21));
    return `${
      isNaN(this.value)
        ? "-"
        : this.value.toLocaleString(undefined, {
            minimumSignificantDigits: sf,
            maximumSignificantDigits: sf
          })
    } ${this.unit}`;
  }

  toJSON() {
    // eslint-disable-next-line no-unused-vars
    let { _value, _unit, ...properties } = this;
    return {
      ...properties,
      value: this.value,
      unit: this.unit,
      type: this.constructor.name
    };
  }

  static significantFigures(n) {
    let s;
    if (n === 0) {
      return 0;
    } else if (typeof n === "number") {
      s = Math.abs(n).toExponential();
    } else if (typeof n === "string" && !Number.isNaN(Number(n))) {
      s = n.replace(/^\+|-/, "").replace(/^0+/, "");
    } else return;

    s = s.replace(".", "");
    const i = s.lastIndexOf("e");
    return i > 0 ? i : s.length;
  }

  static significantValue(val, precision = 1) {
    if (val === 0) return 0;
    const fact = Math.pow(
      10,
      precision - Math.floor(Math.log(val < 0 ? -val : val) / Math.LN10) - 1
    );
    return Math.round(val * fact) / fact;
  }

  static get units() {
    throw "list of available units should be set on constructor.";
  }

  static get baseUnit() {
    return Object.keys(this.units)[Object.values(this.units).indexOf(1)];
  }

  static from({ type, value, unit, ...properties } = {}) {
    if (!type || this.name == type) return new this(value, unit, properties);
  }
}

export class Distance extends Quantity {
  static get units() {
    return {
      m: 1,
      Km: 0.001,
      ft: 3.28084,
      NM: 0.000539957,
      mi: 0.000621371,
      yd: 1.09361
    };
  }
}

export class Altitude extends Distance {
  constructor(value, unit, { reference, ...properties } = {}) {
    if (unit == "FL") reference = "QNE";
    else if (unit && !reference)
      throw `property 'reference' should be given, got '${reference}'`;

    super(value, unit, { ...properties, reference });
  }

  // densityAltitude(temp) {
  //   return new this.constructor(
  //     this.value + 36.576 * (temp - 15),
  //     "m",
  //     "MSL"
  //   );
  // }

  get value() {
    if (this.unit == "FL")
      return this.constructor.significantValue(this._to(this._value), 2);
    else return this._to(this._value);
  }

  set value(val) {
    super.value = val;
  }

  toString() {
    if (this.unit == "FL") {
      return `FL${String(this.value).padStart(3, "0")}`;
    } else if (this.value && this.reference) {
      return `${super.toString()} ${this.reference}`;
    }
  }

  static get references() {
    return ["MSL", "AGL", "QNH", "QNE", "QFE", "WGS84"];
  }

  static get units() {
    return {
      m: 1,
      ft: 3.28084,
      FL: 0.0328084
    };
  }
}

export class Speed extends Quantity {
  static get units() {
    return {
      "m/s": 1,
      "km/h": 3.6,
      "ft/min": 196.85,
      kt: 1.94384,
      mph: 2.23694
    };
  }

  // TAS(altitude, temperature) {}
}

export class Volume extends Quantity {
  static get units() {
    return {
      m3: 1,
      L: 1000,
      "gal US": 264.172,
      "gal Imp": 219.969
    };
  }
}

export class Consumption extends Volume {
  static get references() {
    return ["h", "1000ft", "u"];
  }

  toString(precision) {
    return `${super.toString(precision)}/${this.reference}`;
  }
}

export class Weight extends Quantity {
  static get units() {
    return {
      kg: 1,
      lb: 2.20462,
      t: 0.001,
      "L-avgas": 1 / 0.721,
      "L-JetA1": 1 / 0.8
    };
  }
}

export class Angle extends Quantity {
  static get units() {
    return {
      rad: 1,
      "°": 180 / Math.PI,
      gon: 200 / Math.PI,
      rev: 1 / 2 / Math.PI,
      "₥": 6400 / 2 / Math.PI
    };
  }
}

export class Azimuth extends Angle {
  get value() {
    //FIXME: 0 should give 360
    return this._to((this._value + 2 * Math.PI) % (2 * Math.PI));
  }

  set value(val) {
    super.value = val;
  }
}

export class Bearing extends Angle {
  get value() {
    //FIXME: 180 % should give 180
    return this._to(((this._value + 3 * Math.PI) % (2 * Math.PI)) - Math.PI);
  }

  set value(val) {
    super.value = val;
  }
}

export class Pressure extends Quantity {
  static get units() {
    return {
      Pa: 1,
      hPa: 0.01,
      Psi: 1 / 6894.76,
      atm: 1 / 101325,
      bar: 0.00001,
      inHg: 1 / 3386.389
    };
  }
}

// TODO Spec it !
export class Temperature extends Quantity {
  static get units() {
    return {
      "°C": [1, 0],
      K: [1, 273.15],
      "°F": [9 / 5.32]
    };
  }

  _from(val, unit = this.unit) {
    return Number(val) / this._factor(unit)[0] + this._factor(unit)[1];
  }

  _to(val, unit = this.unit) {
    return (Number(val) - this._factor(unit)[1]) * this._factor(unit)[0];
  }
}
export class Frequency extends Quantity {
  static get units() {
    return {
      Hz: 1,
      KHz: 0.001,
      MHz: 0.00001,
      GHz: 0.00000001
    };
  }
}
