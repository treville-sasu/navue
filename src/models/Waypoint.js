import { Model } from "@/models/Base.js";
import { Angle } from "@/models/Quantities.js";

import { point, round } from "@turf/helpers";
import rhumbDistance from "@turf/rhumb-distance";
import rhumbBearing from "@turf/rhumb-bearing";
import rhumbDestination from "@turf/rhumb-destination";
import truncate from "@turf/truncate";
import greatCircle from "@turf/great-circle";

export class Waypoint extends Model {
  constructor([longitude, latitude, altitude] = [], properties = {}) {
    if (altitude !== undefined) {
      properties.altitudeReference || (properties.altitudeReference = "WGS84");
      altitude = round(Number(altitude));
    }
    super(
      point([round(longitude, 6), round(latitude, 6), altitude], properties)
    );
  }

  set longitude(val) {
    this.geometry.coordinates[0] = round(val, 6);
  }

  set latitude(val) {
    this.geometry.coordinates[1] = round(val, 6);
  }

  get altitude() {
    return this.geometry.coordinates[2];
  }

  set altitude(val) {
    this.geometry.coordinates[2] = round(val);
  }

  get lngLat() {
    return {
      lng: this.geometry.coordinates[0],
      lat: this.geometry.coordinates[1],
    };
  }

  set lngLat({ lng, lat }) {
    this.longitude = lng;
    this.latitude = lat;
  }

  distanceTo(wp) {
    return round(
      rhumbDistance(this.geometry.coordinates, wp.geometry.coordinates, {
        units: "meters",
      })
    );
  }

  bearingTo(wp) {
    return rhumbBearing(this.geometry.coordinates, wp.geometry.coordinates);
  }

  greatCircleTo(wp, properties) {
    let gc = truncate(
      greatCircle(this.geometry.coordinates, wp.geometry.coordinates, {
        npoints: 8,
        properties
      })
    );

    gc.properties = {
      ...gc.properties,
      distance: this.distanceTo(wp),
      bearing: this.bearingTo(wp)
    };

    return gc;
  }

  destinationPoint(distance, heading, properties) {
    const coords = rhumbDestination(
      this.geometry.coordinates,
      distance,
      Angle.wrap180(heading),
      {
        units: "meters",
      }
    ).geometry.coordinates;
    if (this.geometry.coordinates[2]) coords[2] = this.geometry.coordinates[2];
    return new this.constructor(coords, properties);
  }

  toGeoJSON(type) {
    switch (type) {
      case "Position":
        return this.geometry.coordinates;
      case "Point":
        return this.geometry;
      default:
        return {
          ...this,
          type: "Feature",
        };
    }
  }

  static from({
    type,
    geometry,
    coordinates,
    properties = {},
    ...otherProps
  } = {}) {
    if (arguments[0] instanceof this) return arguments[0];

    switch (type) {
      case this.name:
      case "Feature":
        if (geometry && geometry.type == "Point")
          coordinates = geometry.coordinates;
        else
          throw `Invalid data : 'geometry.type' should be 'Point' got '${
            geometry && geometry.type
          }'`;
        break;
      case "Point":
        // coordinates = coordinates;
        break;
      default:
        throw `Invalid data : 'type' should be 'Feature', 'Point' or '${this.name}' got '${type}'`;
    }
    return new this(coordinates, {
      ...otherProps,
      ...properties,
    });
  }

  static fromEvent({ lngLat: { lng, lat } } = {}) {
    return new this([lng, lat]);
  }
}
