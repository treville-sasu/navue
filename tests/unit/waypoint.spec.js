/* eslint-disable jest/no-hooks */
import { Model } from "@/models/Base.js";
import { Waypoint } from "@/models/Waypoint.js";
import { Altitude, Distance, Bearing } from "../../src/models/Quantities";

expect.extend({
  toBeClose(actual, expected, precision = 2) {
    const pass = Math.abs(expected - actual) < Math.pow(10, -precision) / 2;
    if (pass) {
      return {
        message: () => `expected ${actual} not to be close ${expected}`,
        pass: true,
      };
    } else {
      return {
        message: () => `expected ${actual} to be close ${expected}`,
        pass: false,
      };
    }
  },
});

describe("waypoint", () => {
  let wp;
  beforeEach(() => {
    wp = new Waypoint([1.123456789, 1.123456789, 1234.56789], {
      name: "Echo",
      notes: "Some note",
    });
  });
  it("extends Model", () => {
    expect(Waypoint.prototype).toBeInstanceOf(Model);
  });
  it("is instantiated", () => {
    expect(() => new Waypoint()).toThrowError();
    expect(() => new Waypoint("aaaa")).toThrowError();

    expect(wp).toMatchInlineSnapshot(`
      Object {
        "geometry": Object {
          "coordinates": Array [
            1.123457,
            1.123457,
            1235,
          ],
          "type": "Point",
        },
        "properties": Object {
          "altitudeReference": "WGS84",
          "name": "Echo",
          "notes": "Some note",
        },
        "type": "Waypoint",
      }
    `);
  });
  // describe("getter & setter", () => {

  it("handles longitude, latitude & lngLat (with rounding, 0.000001Â°)", () => {
    wp.longitude = 2.123456789;
    expect(wp.geometry.coordinates[0]).toBe(2.123457);
    wp.latitude = 2.123456789;
    expect(wp.geometry.coordinates[1]).toBe(2.123457);

    wp.lngLat = { lat: 1.123456789, lng: 1.123456789 };
    expect(wp.geometry).toMatchInlineSnapshot(`
      Object {
        "coordinates": Array [
          1.123457,
          1.123457,
          1235,
        ],
        "type": "Point",
      }
    `);
    expect(wp.lngLat).toMatchInlineSnapshot(`
      Object {
        "lat": 1.123457,
        "lng": 1.123457,
      }
    `);
  });

  it("handles altitude (with rounding, 1m)", () => {
    wp.altitude = 1.123456789;
    expect(wp.geometry.coordinates[2]).toBe(1);
    expect(wp.altitude).toBe(wp.geometry.coordinates[2]);
  });

  it("get distance to other", () => {
    expect(wp.distanceTo(new Waypoint([1, 1]))).toMatchInlineSnapshot(`19412`);
  });
  it("get a bearing to other", () => {
    expect(
      new Waypoint([0, 0]).bearingTo(new Waypoint([0, 1]))
    ).toMatchInlineSnapshot(`0`);
  });
  it("get destination waypoint from distance and bearing", () => {
    expect(
      new Waypoint([0, 0, 1]).destinationPoint(
        new Distance(157249),
        new Bearing(0)
      )
    ).toMatchSnapshot();
    expect(new Waypoint([0, 0]).destinationPoint(157249, 0)).toMatchObject(
      new Waypoint([0, 0]).destinationPoint(157249, 0)
    );
  });
  it.todo("return a Great Circle");

  it("is exported as JSON object", () => {
    expect(wp.toJSON()).toMatchSnapshot();
  });

  it("is exported as GeoJSON Position, or Point & Feature", () => {
    expect(wp.toGeoJSON("Position")).toMatchSnapshot();
    expect(wp.toGeoJSON("Point")).toMatchSnapshot();
    expect(wp.toGeoJSON()).toMatchSnapshot();
  });

  it("is imported from GeoJSON Position, Point & Feature, or plain Waypoint object", () => {
    expect(Waypoint.from(wp.toJSON())).toMatchObject(wp);
    expect(Waypoint.from(wp.toGeoJSON("Point"))).toMatchObject(
      new Waypoint(wp.geometry.coordinates)
    );
    expect(Waypoint.from(wp.toGeoJSON())).toMatchObject(wp);
    expect(() => Waypoint.from({ type: "Polyline" })).toThrow(
      "Invalid data : 'type' should be 'Feature', 'Point' or 'Waypoint' got 'Polyline'"
    );
  });
  it.skip("is imported from (mapbox), event", () => {
    expect(
      Waypoint.fromEvent({
        type: "click",
        lngLat: { lng: 1.123456789, lat: 1.123456789 },
      })
    ).toMatchObject(wp);
  });
});
