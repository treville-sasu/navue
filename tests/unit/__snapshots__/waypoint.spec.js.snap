// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`waypoint get destination waypoint from distance and bearing 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      1.414172,
      1,
    ],
    "type": "Point",
  },
  "properties": Object {
    "altitudeReference": "WGS84",
  },
  "type": "Waypoint",
}
`;

exports[`waypoint is exported as GeoJSON Position, or Point & Feature 1`] = `
Array [
  1.123457,
  1.123457,
  1235,
]
`;

exports[`waypoint is exported as GeoJSON Position, or Point & Feature 2`] = `
Object {
  "coordinates": Array [
    1.123457,
    1.123457,
    1235,
  ],
  "type": "Point",
}
`;

exports[`waypoint is exported as GeoJSON Position, or Point & Feature 3`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      1.123457,
      1.123457,
      1235,
    ],
    "type": "Point",
  },
  "properties": Object {
    "altitudeReference": "WGS84",
    "name": "Echo",
    "notes": "Some note",
  },
  "type": "Feature",
}
`;

exports[`waypoint is exported as JSON object 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      1.123457,
      1.123457,
      1235,
    ],
    "type": "Point",
  },
  "properties": Object {
    "altitudeReference": "WGS84",
    "name": "Echo",
    "notes": "Some note",
  },
  "type": "Waypoint",
}
`;
